[DDW-716] Implement static screens for signing dApp interaction transactions

daniloprates opened the pull request:
This PR implements static screens for signing dApp interaction transactions.

## Todos

- [x] dApp transaction request dialog
- [x] Transaction confirmation Assets list component
- [x] Missing token
- [x] Add wallet dropdown option
- [x] Togglable content
- [x] Copy + translation

## Screenshots

![screenshot-app zeplin io-2021 08 12-17_52_11](https://user-images.githubusercontent.com/1504716/129268600-baa9fad6-8243-45fe-98ef-adbb42755dba.png)


---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1628801324059600)
- [ ] Test

---

## Review Checklist

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

gabriela-ponce commented:
@daniloprates A few comments for this dialog:

- When the amount of the token is not enough, the red tooltip overlaps the other tooltips for that element. Check [this](https://drive.google.com/file/d/1J7n4MGaQgSZMbd2QfiTzjtn4o7zL4bbn/view?usp=sharing) video. Can this be improved? Maybe changing the orientation of the red tooltip?
- There are a few translations missing in Japanese. Check screenshots [1](https://drive.google.com/file/d/1fpV6attkOPYF4wTyfHwGsn8DOhVRBcrU/view?usp=sharing), [2](https://drive.google.com/file/d/1yxNk1-8vg-w0qMCZ7BeL7ZtP-LjNvxRz/view?usp=sharing) and [3](https://drive.google.com/file/d/1cRY8i3GAm4kHOrLR94ldZNnPI0Cu1-6n/view?usp=sharing)
- When there is no enough ADA, there is [no indication on the dialog](https://drive.google.com/file/d/12w-ry1hvFkkWhA29h2YDMVFZT4gBkgF3/view?usp=sharing). Shouldn't the dropdown be outlined in red as on [this](https://app.zeplin.io/project/57acfaa7436a750316161a8b/screen/60ebfaeec23b321391ad1e21) design? Same with the scenario when [the token isn't enough.](https://drive.google.com/file/d/1slgG4fdcKHewCZN4Hkqx0we8BjWN38if/view?usp=sharing)

daniloprates commented:
@gabriela-ponce thanks! I've fixed the items:


> * When the amount of the token is not enough, the red tooltip overlaps the other tooltips for that element. Check [this](https://drive.google.com/file/d/1J7n4MGaQgSZMbd2QfiTzjtn4o7zL4bbn/view?usp=sharing) video. Can this be improved? Maybe changing the orientation of the red tooltip?

Fixed for [both](https://user-images.githubusercontent.com/1504716/129460488-1c44e464-1fdf-4e94-a1fb-30368ec5507b.png) [cases](https://user-images.githubusercontent.com/1504716/129460490-6e22fb43-7aba-4457-b1f9-0354672a2340.png). Please notice, in case the dialog height is not tall enough, it will [open upwards](https://user-images.githubusercontent.com/1504716/129460521-d5e26b4c-e56f-4260-9918-71e361e252ce.png) and there's nothing we can do.

> * There are a few translations missing in Japanese. Check screenshots [1](https://drive.google.com/file/d/1fpV6attkOPYF4wTyfHwGsn8DOhVRBcrU/view?usp=sharing), [2](https://drive.google.com/file/d/1yxNk1-8vg-w0qMCZ7BeL7ZtP-LjNvxRz/view?usp=sharing) and [3](https://drive.google.com/file/d/1cRY8i3GAm4kHOrLR94ldZNnPI0Cu1-6n/view?usp=sharing)

All fixed ツ

> * When there is no enough ADA, there is [no indication on the dialog](https://drive.google.com/file/d/12w-ry1hvFkkWhA29h2YDMVFZT4gBkgF3/view?usp=sharing). Shouldn't the dropdown be outlined in red as on [this](https://app.zeplin.io/project/57acfaa7436a750316161a8b/screen/60ebfaeec23b321391ad1e21) design? Same with the scenario when [the token isn't enough.](https://drive.google.com/file/d/1slgG4fdcKHewCZN4Hkqx0we8BjWN38if/view?usp=sharing)

It's [done](https://user-images.githubusercontent.com/1504716/129460551-70a9c23a-3621-4115-996e-2be2a184b876.png). I'll reach for proper copy & translation on Monday.

dmitrii-gaico commented:
@daniloprates Hey Danilo. 

I have a question regarding latest fix for scenario when wallet has not enough ADA. If this happens shouldn't we disable [Continue button](https://drive.google.com/file/d/1bP6A-W1W4Q7rHb7crJAruiwFY-tDT7_E/view?usp=sharing) ? If wallet has not enough ADA then it can't proceed.

daniloprates commented:
> @daniloprates Hey Danilo.
> 
> I have a question regarding latest fix for scenario when wallet has not enough ADA. If this happens shouldn't we disable [Continue button](https://drive.google.com/file/d/1bP6A-W1W4Q7rHb7crJAruiwFY-tDT7_E/view?usp=sharing) ? If wallet has not enough ADA then it can't proceed.

Nice catch, @dmitrii-gaico. Fixed it.

gabriela-ponce commented:
@daniloprates All the fixes look great, nice job!

[Here is a screenshot](https://drive.google.com/file/d/18_t2V55YP3Zrwov4gxuaIzgQy_D7Ffwj/view?usp=sharing) of some translations in Japanese that are still missing (I know the first one is still pending, mentioning it for tracking purposes).

miorsufianiohk commented:
Hi @daniloprates ,

Tested on 18800.

I am not sure if this is relevant or not but I'm including this for consistency reason. I noticed that the displayed index of the token starts from position "0" [screenshot 1](https://drive.google.com/file/d/1YIqKbUHk8AoVj_WI1ZidVKZH9loRa_Zi/view?usp=sharing), in the webApp when sending multiple tokens, it starts from position "1" [screenshot 2](https://drive.google.com/file/d/1QUFutbgDRCj8IZdXbf52EZidmpdCQGLS/view?usp=sharing).

daniloprates commented:
@gabriela-ponce @miorsufianiohk thanks for the reports. I've added the missing Japanese translation and increased the Token number, so it starts at 1

miorsufianiohk reviewed (approved):
LGTM. Great job @daniloprates  👍 . Tested on 18809.

alexander-rukin commented:
<img width="677" alt="image" src="https://user-images.githubusercontent.com/2539345/130206601-b61a85d7-7791-448f-85e0-af5cb7c1ee89.png">
something wrong with scroll

alexander-rukin commented:
<img width="469" alt="image" src="https://user-images.githubusercontent.com/2539345/130206935-45da43fa-f958-4606-9d26-15f61cbbd4de.png">
confused a bit with error message and popup shouldn't be red I suppose

alexander-rukin commented:
<img width="592" alt="image" src="https://user-images.githubusercontent.com/2539345/130207156-322fc954-80b4-4627-bb59-02f694bae76d.png">
<img width="600" alt="image" src="https://user-images.githubusercontent.com/2539345/130207186-6bb0d185-aaee-4610-9409-33403ab7a319.png">
Errors are about almost the same but one has explanations as a separate sentence, other in brackets

alexander-rukin commented:
<img width="975" alt="image" src="https://user-images.githubusercontent.com/2539345/130207576-6f27fe52-5b7c-42ce-bbe6-b23d91a63cfe.png">
looks like this button should be moved to the top (1px)

alexander-rukin commented:
<img width="308" alt="image" src="https://user-images.githubusercontent.com/2539345/130207882-c0b66444-a574-4c01-82e0-6179cbed272f.png">
missing spacing between elements

alexander-rukin commented:
<img width="627" alt="image" src="https://user-images.githubusercontent.com/2539345/130208073-6f2f90ca-aa64-4cb3-bd18-9b64694af2e1.png">
<img width="287" alt="image" src="https://user-images.githubusercontent.com/2539345/130208394-748e5de9-080e-47a5-96ce-86055927857b.png">
with 20px (instead of 25px) it looks better

alexander-rukin commented:
<img width="622" alt="image" src="https://user-images.githubusercontent.com/2539345/130208528-8d0f136c-468c-45e4-b811-3e46d01043a5.png">
Initially, I planned wallet input should be red in any error case, as it's the source of a problem :D any thoughts?

danielmain commented on source/renderer/app/components/notifications/NotificationActions.js:
```diff
@@ -0,0 +1,66 @@
+  const componentStyles = classnames([styles.component]);
+  return (
+    <div className={componentStyles}>
+      {map(actions, (action, key) => {
```
Maybe destructuring in line 33 saves you code?  like: "{map(actions, ({className, primary, autoFocus, label, onClick, disabled ....}, key) => {

daniloprates replied:
Done

danielmain reviewed (changes requested):
Great Job Danilo. I just added some small semantic corrections as review

dmitrii-gaico commented:
Not sure if someone already reported a typo so here it is :
![image](https://user-images.githubusercontent.com/83477982/130243060-575dfefc-b85c-4f04-a6c9-9ac9b0cec46b.png)

daniloprates commented:
> Initially, I planned wallet input should be red in any error case, as it's the source of a problem :D any thoughts?

@alexander-rukin It can be confusing because hovering the wallets dropdown won't give you any feedback

![image](https://user-images.githubusercontent.com/1504716/130434597-3182d20c-d004-473e-bdb0-d4b17bbbf597.png)

alexander-rukin commented:
> > Initially, I planned wallet input should be red in any error case, as it's the source of a problem :D any thoughts?
> 
> @alexander-rukin It can be confusing because hovering the wallets dropdown won't give you any feedback

it's ok, we highlight error in TX details (we can also consider activating those error dropdown pointing error line while hovering input too) 

oh... btw here we should highlight ada row (as this is cause of an issue) 

<img width="657" alt="image" src="https://user-images.githubusercontent.com/2539345/130439852-12654565-8f18-406c-acd3-97d80a5daa48.png">

daniloprates commented on source/renderer/app/components/assets/AssetsTransactionConfirmation.js:
```diff
@@ -0,0 +1,53 @@
+  getAssetByUniqueId: Function,
+};
+
+const AssetsTransactionConfirmation = observer((props: Props) => {
```
:)

daniloprates commented on source/renderer/app/components/dapp/DappTransactionRequest.js:
```diff
@@ -0,0 +1,238 @@
+    wallets,
+  } = props;
+  const walletsDropdownError =
+    selectedWallet &&
```
I had missed this feature in JS 😱 thanks for taking me out of darkness

daniloprates commented:
@alexander-rukin @danielmain @dmitrii-gaico thanks for your review and great suggestions. They're done, except the copy changes, which [I've asked](https://input-output-rnd.slack.com/archives/CBECZ2K47/p1629732874016500?thread_ts=1628692075.009300&cid=CBECZ2K47) Nyamh to have a look.

daniloprates commented on source/renderer/app/components/assets/Asset.js:
```diff
@@ -84,6 +84,7 @@ type Props = {
   // In case it's not possible to calculate the container width
   // this props defines after how many characters the `metadata.name` text will cut off
   metadataNameChars?: number,
+  error?: boolean,
```
Done

daniloprates commented on source/renderer/app/components/assets/AssetContent.js:
```diff
@@ -75,6 +75,7 @@ type Props = {
   highlightFingerprint?: boolean,
   className?: string,
   intl: intlShape.isRequired,
+  error?: boolean,
```
Done

daniloprates commented on source/renderer/app/components/dapp/DappTransactionRequest.js:
```diff
@@ -0,0 +1,238 @@
+    triggeredFrom,
+    wallets,
+  } = props;
+  const walletsDropdownError =
```
Done

daniloprates commented on source/renderer/app/components/dapp/DappTransactionRequest.js:
```diff
@@ -0,0 +1,238 @@
+          numberOfStakePools={100}
+          onChange={onSelectWallet}
+          placeholder={intl.formatMessage(messages.walletsDropdownPlaceholder)}
+          value={selectedWallet ? selectedWallet.id : null}
```
Done

daniloprates commented on source/renderer/app/components/notifications/Notification.js:
```diff
@@ -79,6 +85,8 @@ export default class Notification extends Component<Props> {
 
             <div className={messageStyles}>{children}</div>
 
+            {actions && <NotificationActions actions={actions} />}
```
Done

alexander-rukin commented:
<img width="550" alt="image" src="https://user-images.githubusercontent.com/2539345/130631845-0658bbf6-a1b5-4fd2-90bb-92a2dc6fe964.png">
"x" is not centered in the notification

alexander-rukin commented:
<img width="620" alt="image" src="https://user-images.githubusercontent.com/2539345/130632227-a3132877-bdcb-491c-a898-d1eea44a0d5f.png">

you probably got me wrong, I suggested to make ada row red in TX details, not the dropdown

<img width="604" alt="image" src="https://user-images.githubusercontent.com/2539345/130632358-c6a3ef23-f479-4ae9-baa1-c4681e889640.png">


it also makes other balances red

alexander-rukin commented:
<img width="97" alt="image" src="https://user-images.githubusercontent.com/2539345/130632794-eaec5ba0-e6f3-444a-805c-21ba0bff22ad.png">

isn't it possible to have a shift like in designs?

<img width="90" alt="image" src="https://user-images.githubusercontent.com/2539345/130632897-671a2dd0-9789-4cd5-9a1e-98438246335d.png">

p.s. shouldn't it be "Ada" here and not "ADA"?)

alexander-rukin commented:
<img width="244" alt="image" src="https://user-images.githubusercontent.com/2539345/130633083-e3848506-a955-4be6-accb-2668d346938f.png">

distance looks too big 

<img width="286" alt="image" src="https://user-images.githubusercontent.com/2539345/130633119-a1d53dd6-1eca-4be4-9edb-8a2f76707eca.png">

here design reference

daniloprates commented:
> <img alt="image" width="97" src="https://user-images.githubusercontent.com/2539345/130632794-eaec5ba0-e6f3-444a-805c-21ba0bff22ad.png">
> 
> isn't it possible to have a shift like in designs?
> 
> <img alt="image" width="90" src="https://user-images.githubusercontent.com/2539345/130632897-671a2dd0-9789-4cd5-9a1e-98438246335d.png">
> 
> p.s. shouldn't it be "Ada" here and not "ADA"?)

@alexander-rukin what do you mean by "shift"? (For amounts we use the ticker, ADA)

alexander-rukin commented:
> > <img alt="image" width="97" src="https://user-images.githubusercontent.com/2539345/130632794-eaec5ba0-e6f3-444a-805c-21ba0bff22ad.png">
> > isn't it possible to have a shift like in designs?
> > <img alt="image" width="90" src="https://user-images.githubusercontent.com/2539345/130632897-671a2dd0-9789-4cd5-9a1e-98438246335d.png">
> > p.s. shouldn't it be "Ada" here and not "ADA"?)
> 
> @alexander-rukin what do you mean by "shift"? (For amounts we use the ticker, ADA)
Yes, for amounts "ADA", but for the line name I guess it should be still "ada" or "Ada" (if at the beginning of sentence) 
As for the shift -> look at the vertical line position

daniloprates commented:
> > > <img alt="image" width="97" src="https://user-images.githubusercontent.com/2539345/130632794-eaec5ba0-e6f3-444a-805c-21ba0bff22ad.png">
> > > isn't it possible to have a shift like in designs?
> > > <img alt="image" width="90" src="https://user-images.githubusercontent.com/2539345/130632897-671a2dd0-9789-4cd5-9a1e-98438246335d.png">
> > > p.s. shouldn't it be "Ada" here and not "ADA"?)
> > 
> > 
> > @alexander-rukin what do you mean by "shift"? (For amounts we use the ticker, ADA)
> 
> Yes, for amounts "ADA", but for the line name I guess it should be still "ada" or "Ada" (if at the beginning of sentence)
> 
> As for the shift -> look at the vertical line position

@alexander-rukin done!

alexander-rukin commented:
<img width="224" alt="image" src="https://user-images.githubusercontent.com/2539345/131534663-e624e786-dfde-4c8d-9356-c0463cc119fc.png">
in the collapsed view it still lacks margin

alexander-rukin commented:
<img width="97" alt="image" src="https://user-images.githubusercontent.com/2539345/131534925-96bdd3ff-40d4-40b6-91ca-e16271ab3b68.png">
hover effect here should be same as for the normal icon (increase opacity on hover)

alexander-rukin commented:
<img width="750" alt="image" src="https://user-images.githubusercontent.com/2539345/131537889-aff7fb25-424f-4021-ae8c-909b13a6dc37.png">
scrollbars still look like this

alexander-rukin commented:
<img width="182" alt="image" src="https://user-images.githubusercontent.com/2539345/131539486-39de71a8-ab28-4244-97fb-d1bc0e3f2b4d.png">
not sure what caused this, but right end of a content is shifted to the left. Content should end where button ends

daniloprates commented:
@alexander-rukin done, except:

> <img alt="image" width="750" src="https://user-images.githubusercontent.com/2539345/131537889-aff7fb25-424f-4021-ae8c-909b13a6dc37.png">
> 
> scrollbars still look like this


Which browser are you using? It needs to be Chrome

> <img alt="image" width="182" src="https://user-images.githubusercontent.com/2539345/131539486-39de71a8-ab28-4244-97fb-d1bc0e3f2b4d.png">
> 
> not sure what caused this, but right end of a content is shifted to the left. Content should end where button ends

Didn't understand.


This is how it looks to me:






![Screen Shot 2021-08-31 at 16 17 12](https://user-images.githubusercontent.com/1504716/131562713-52bc255e-fd6f-4190-adbd-66b7854c1b98.png)

alexander-rukin commented:
> Which browser are you using? It needs to be Chrome

I use Chrome Version 92.0.4515.159 (Official Build) (x86_64)

gabriela-ponce commented:
@daniloprates The fixes look good to me, but the issue that @alexander-rukin mentioned about the scroll is still visible (also using Chrome).
- The scroll bar is only visible when scrolling with the mouse (can't click on it if you don't scroll first). Also, on hover, it looks bigger. Check [this](https://drive.google.com/file/d/1paq6OBi2jaJ9SYdjwpnv5Tt1mYt-p9nI/view?usp=sharing) video.

alexander-rukin commented:
<img width="607" alt="image" src="https://user-images.githubusercontent.com/2539345/131675800-cb04f5a4-c2b7-49ac-bca5-a87a22d1d670.png">
btw I'd expect to see this error message when hovering bottom line in TX (same way as in previous cases)

daniloprates commented:
> @daniloprates The fixes look good to me, but the issue that @alexander-rukin mentioned about the scroll is still visible (also using Chrome).
> 
> * The scroll bar is only visible when scrolling with the mouse (can't click on it if you don't scroll first). Also, on hover, it looks bigger. Check [this](https://drive.google.com/file/d/1paq6OBi2jaJ9SYdjwpnv5Tt1mYt-p9nI/view?usp=sharing) video.

This is a default behavior for mac. You can change it globally [here](https://user-images.githubusercontent.com/1504716/131686014-c61bb589-0efd-4222-8428-49ca59ef1171.png).

@gabriela-ponce @alexander-rukin I've fixed the different color for the data boxes. They're not as the main dialog. We may need to change the color for all dialog scrollbars, but I prefer to do that in a separated task, testing in the actual app, not only Storybook.

daniloprates commented:
@alexander-rukin added the Ada error here as well 👇

![image](https://user-images.githubusercontent.com/1504716/131697470-e54abad1-480f-4d8d-b299-b74e7ed5cc17.png)

alexander-rukin commented:
for some reason “computed” line-height here (Ada line)  is 21px -> with 22px it looks better

![image](https://user-images.githubusercontent.com/2539345/131705745-c51b3c61-060e-4b38-99a8-e56da5699365.png)

---
all line-heights here 21px
I made all of them 22, so line-height = 22px, box-height = 22px
here on a screenshot look at the alignment:
line - 21px
red box - 22px, but shifted to the top
violet box - 21px and aligned with a text
p.s. left violet box just to see that red is bigger

![image](https://user-images.githubusercontent.com/2539345/131705868-dc05b8ef-98e8-4260-9062-11355953a917.png)


---
btw receiver line-height seems to be 22, as it’s 44 in total ))

![image](https://user-images.githubusercontent.com/2539345/131705896-3eb411cf-e573-44d7-a85e-0ca7816699f0.png)
![image](https://user-images.githubusercontent.com/2539345/131705921-a1f59883-d39b-4a42-8850-38fc01d6416e.png)

daniloprates commented:
> for some reason “computed” line-height here (Ada line) is 21px -> with 22px it looks better

> all line-heights here 21px
> I made all of them 22, so line-height = 22px, box-height = 22px

@alexander-rukin done

![Screen Shot 2021-09-01 at 17 13 05](https://user-images.githubusercontent.com/1504716/131738489-a6513700-dd35-4f1b-a7c4-76f82bbe967c.png)
![Screen Shot 2021-09-01 at 15 37 51](https://user-images.githubusercontent.com/1504716/131738499-96bf0181-05b9-4eb5-a5f0-a607ed77b445.png)

alexander-rukin commented:
> @alexander-rukin added the Ada error here as well 👇
> 
> ![image](https://user-images.githubusercontent.com/1504716/131697470-e54abad1-480f-4d8d-b299-b74e7ed5cc17.png)

did you push this?

alexander-rukin commented:
<img width="198" alt="image" src="https://user-images.githubusercontent.com/2539345/131799919-0d981cb2-cd48-497d-bc28-7d89bdd4e29c.png">

Just noticed that "ADA" suffix has wrong styles - see https://zpl.io/2yDLwD9

alexander-rukin commented:
<img width="711" alt="image" src="https://user-images.githubusercontent.com/2539345/131801414-710a11b3-29b3-4943-bd8f-d80f4c5e482b.png">
I can see that scroll bar is not centered on your screenshot 😓

alexander-rukin reviewed (approved):
All good now. @daniloprates you are a star ⭐

gabriela-ponce commented:
@daniloprates The fixes look good. Just a little comment about [this tooltip](https://user-images.githubusercontent.com/1504716/131697470-e54abad1-480f-4d8d-b299-b74e7ed5cc17.png). Wouldn't it look better to be displayed over the value (100.00000 in this case)? 
I know the current implementation is also how it's done for the tokens, but I think it looks a little odd, as it's not clear what is pointing to.

daniloprates commented:
> @daniloprates The fixes look good. Just a little comment about [this tooltip](https://user-images.githubusercontent.com/1504716/131697470-e54abad1-480f-4d8d-b299-b74e7ed5cc17.png). Wouldn't it look better to be displayed over the value (100.00000 in this case)?
> I know the current implementation is also how it's done for the tokens, but I think it looks a little odd, as it's not clear what is pointing to.

@gabriela-ponce nice suggestion. The only issue is that, because the Ada Amount element is narrow, it can be opened from the top, as it looks like [this](https://user-images.githubusercontent.com/1504716/131914686-fd98b40d-7e5c-49f6-8969-eb868634878c.png). From the left it works [fine](https://user-images.githubusercontent.com/1504716/131914744-5047e546-f367-4775-8a84-ab46c09cfb73.png). @Alexander-rukin what do you think?

alexander-rukin commented:
> @gabriela-ponce nice suggestion. The only issue is that, because the Ada Amount element is narrow, it can be opened from the top, as it looks like [this](https://user-images.githubusercontent.com/1504716/131914686-fd98b40d-7e5c-49f6-8969-eb868634878c.png). From the left it works [fine](https://user-images.githubusercontent.com/1504716/131914744-5047e546-f367-4775-8a84-ab46c09cfb73.png). @alexander-rukin what do you think?

I thought we can make the tooltip globally positioned. So it will just go out of bounds for dialog. 

As for now, the tooltip points to the whole row which is an error one. If we want to change position and point exact error part, we will need to do this for every case here. Point amount when it's an amount problem. Point token box when "no such token in wallet" error. But still the whole row should be highlighted in red

daniloprates commented:
> > @gabriela-ponce nice suggestion. The only issue is that, because the Ada Amount element is narrow, it can be opened from the top, as it looks like [this](https://user-images.githubusercontent.com/1504716/131914686-fd98b40d-7e5c-49f6-8969-eb868634878c.png). From the left it works [fine](https://user-images.githubusercontent.com/1504716/131914744-5047e546-f367-4775-8a84-ab46c09cfb73.png). @alexander-rukin what do you think?
> 
> I thought we can make the tooltip globally positioned. So it will just go out of bounds for dialog.
> 
> As for now, the tooltip points to the whole row which is an error one. If we want to change position and point exact error part, we will need to do this for every case here. Point amount when it's an amount problem. Point token box when "no such token in wallet" error. But still the whole row should be highlighted in red

@gabriela-ponce @alexander-rukin reverted the side positioning

danielmain commented on source/renderer/app/components/dapp/DappTransactionRequest.js:
```diff
@@ -0,0 +1,254 @@
+    false
+  );
+  const [isMetadataVisible, toggleMetadata] = useState<boolean>(false);
+  const getToggleLabel = (isVisible: boolean) =>
```
Using useCallback you avoid re-defining the function everytime. Additionally its get memoized `const getToggleLabel = useCallback((isVisible: boolean) => { ... }))`

danielmain commented on source/renderer/app/components/dapp/DappTransactionRequest.js:
```diff
@@ -0,0 +1,254 @@
+  } = props;
+  let hasAmountError = false;
+  const hasTokenError = assets.reduce((result, token, index) => {
+    if (!selectedWallet) return false;
```
Also here, this reduce is called everytime. I would use useCallback or useMemo.  https://reactgo.com/react-usememo-example/

danielmain reviewed (changes requested):
I like how you write clean code :) I just added some minor performance Issues when working with functional components. But overall very good!

daniloprates commented:
> I like how you write clean code :) I just added some minor performance Issues when working with functional components. But overall very good!

Thanks, @danielmain. It's done
